packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -944,24 +924,40 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 	}
 
 	public async getPendingBlobs(waitBlobsToAttach?: boolean): Promise<IPendingBlobs | undefined> {
-		if (waitBlobsToAttach) {
-			await this.stashPendingBlobs();
-		}
-
 		if (this.pendingBlobs.size === 0) {
 			return;
 		}
-
 		const blobs = {};
-		for (const [key, entry] of this.pendingBlobs) {
-			blobs[key] = {
-				blob: bufferToString(entry.blob, "base64"),
-				storageId: entry.storageId,
-				attached: entry.attached,
-				acked: entry.acked,
-				minTTLInSeconds: entry.minTTLInSeconds,
-				uploadTime: entry.uploadTime,
-			};
+		const localIds = new Set<PendingBlob>();
+		while (localIds.size < this.pendingBlobs.size) {
+			for (const [id, entry] of this.pendingBlobs) {
+				if (!localIds.has(entry)) {
+					localIds.add(entry);
+					if (waitBlobsToAttach) {
+						if (!entry.opsent) {
+							this.sendBlobAttachOp(id, entry.storageId);
+						}
+						entry.handleP.resolve(this.getBlobHandle(id));
+						if (!entry.attached) {
+							await new Promise<void>((resolve) => {
+								this.on("blobAttached", (attachedEntry) => {
```

Comment:
you need to remove (off) the listener when the promise is resolved. also what about the abort signal?